lib/services/parser.go

```diff
@@ -518,3 +522,142 @@ func newParserForIdentifierSubcondition(ctx RuleContext, identifier string) (pre
+		return func() bool {
+			return areOperandsEqual(a, b)
+		}
+	}
+	neqPred := func(a interface{}, b interface{}) predicate.BoolPredicate {
+		return func() bool {
+			return !areOperandsEqual(a, b)
+		}
+	}
+
+	p, err := predicate.NewParser(predicate.Def{
+		Operators: predicate.Operators{
+			AND: predicate.And,
+			OR:  predicate.Or,
+			NOT: predicate.Not,
+			EQ:  eqPred,
+			NEQ: neqPred,
+		},
+		Functions: map[string]interface{}{
+			"equals": eqPred,
+			// search allows fuzzy matching against select field values.
+			"search": func(searchVals ...string) predicate.BoolPredicate {
+				return func() bool {
+					return resource.MatchSearch(searchVals)
+				}
+			},
+			// exists checks for an existence of a label by checking
+			// if a key exists. Label value are unchecked.
+			"exists": func(l label) predicate.BoolPredicate {
+				return func() bool {
+					return len(l.key) > 0
+				}
+			},
+		},
+		GetIdentifier: func(fields []string) (interface{}, error) {
+			switch fields[0] {
+			case ResourceLabelsIdentifier:
+				combinedLabels := resource.GetAllLabels()
+				switch {
+				// Field length of 1 means the user is using
+				// an index expression ie: labels["env"], which the
+				// parser will expect a map for lookup in `GetProperty`.
+				case len(fields) == 1:
+					return labels(combinedLabels), nil
+				case len(fields) > 2:
+					return nil, trace.BadParameter("only two fields are supported with identifier %q, got %d: %v", ResourceLabelsIdentifier, len(fields), fields)
+				default:
+					key := fields[1]
+					val, ok := combinedLabels[key]
+					if ok {
+						return label{key: key, value: val}, nil
+					}
+					return label{}, nil
+				}
+
+			case ResourceNameIdentifier:
+				switch {
+				case len(fields) > 1:
+					return nil, trace.BadParameter("only one field are supported with identifier %q, got %d: %v", ResourceNameIdentifier, len(fields), fields)
+				default:
+					return resource.GetName(), nil
+				}
+			case ResourceIdentifier:
+				return predicate.GetFieldByTag(resource, teleport.JSON, fields[1:])
+			default:
+				return nil, trace.NotFound("%v is not defined", strings.Join(fields, "."))
+			}
+		},
+		GetProperty: func(mapVal, keyVal interface{}) (interface{}, error) {
+			m, ok := mapVal.(labels)
+			if !ok {
+				return GetStringMapValue(mapVal, keyVal)
+			}
+
+			key, ok := keyVal.(string)
+			if !ok {
+				return nil, trace.BadParameter("only string keys are supported")
+			}
+
+			val, ok := m[key]
+			if ok {
+				return label{key: key, value: val}, nil
+			}
+			return label{}, nil
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return boolPredicateParser{Parser: p}, nil
+}
+
+// areOperandsEqual is similar to `predicate.Equals`, but
+// adds a type label switch case, and returns a bool instead of a
+// bool predicate function to be able to use it more flexibly.
+func areOperandsEqual(a interface{}, b interface{}) bool {
```

Comment:
this is similar to https://github.com/gravitational/predicate/blob/master/lib.go#L57, i am wondering if it's more proper to do this in the predicate library: create a `areOperandsEqual` and have `Equals` return a closure calling `areOperandsEqual`. And then here, we can extend the switch in `areOperandsEqual` with a label case.

Reply:
I'm not so sure that should belong in predicate. I think we can instead rely on `pedicate.Equals` here. What do you think about this change?
```suggestion
func areOperandsEqual(a interface{}, b interface{}) bool {
	switch aval := a.(type) {
	case label:
		bval, ok := b.(string)
		return ok && aval.value == bval
	default:
		return predicate.Equals(a, b)()
	}
}
```

Reply:
i think it's legit, ty!

Reply:
Perhaps make `areOperandsEqual` return a `predicate.BoolPredicate`?

Reply:
@espadolini I needed it to return a bool so I could use it flexibly like below. 
```go
        eqPred := func(a interface{}, b interface{}) predicate.BoolPredicate {
		return func() bool {
			return areOperandsEqual(a, b)
		}
	}
	neqPred := func(a interface{}, b interface{}) predicate.BoolPredicate {
		return func() bool {
			return !areOperandsEqual(a, b)
		}
	}
```

Or we could use same technique, if we want it to return a `predicate.BoolPredicate`:
```go
        eqPred := areOperandsEqual // don't need it in a variable anymore but for example sake
	neqPred := func(a interface{}, b interface{}) predicate.BoolPredicate {
		return func() bool {
			return !areOperandsEqual(a, b)()
		}
	}
```
Is this what you were thinking of?

Reply:
```go
neqPred := areOperandsEqual.Not()
```

would work, thankfully it's from `v1.1.0` which is what we're using already.

Reply:
ty @espadolini https://github.com/gravitational/teleport/pull/10008/commits/539f700fe0530b8ddd791ece330a69155da9e46a